#[1번 문제]
클래스와 모듈의 공통점과 차이점에 대해 설명하시오.

=> 이름 공간이란, 이름이 존재하는 장소로, 파이썬은 실행 시간에 이름들을 적절한 이름 공간에 넣어 관리합니다.
모듈과 클래스는 독립된 이름 공간(스코프)를 제공한다는 공통점이 있습니다. 즉 변수와 함수를 한곳으로 묶어 사용하는 것 입니다. 반면
모듈은 파일 단위로 이름 공간을 구성하고 코드를 묶는 것으로 이것이 추후 재사용하는 단위이며, 클래스는 클래스 영역 내에 이름 공간을 구성하며, 객체를 만들어 사용한다는 차이점이 있습니다.

#[2번 문제]
다형성에 대해 설명하고 다형성을 보여주는 자신만의 파이썬 코드 예제를 제시하시오.

In [15]:
class Ironman:
    def jarvis(self):
        print "The truth is.... I am Ironman"
        
class Mark1(Ironman):
    def jarvis(self):
        print "Mark1"
        
class Mark7(Ironman):
    def jarvis(self):
        print 'Mark7'
        
class Mark42(Ironman):
    pass

for marks in (Mark1(), Mark7(), Mark42()):
    marks.jarvis()

Mark1
Mark7
The truth is.... I am Ironman


* 다형성이란 상속 관계 내의 다른 클래스들의 인스턴스들이 같은 멤버 함수 호출에 대해 각각 다르게 반응하도록 하는 기능입니다.
* 즉, a와 b의 객체 형에 따라 a+b의 +연산자 행동 방식이 변경되는 것과 같은 것 입니다.
* 위의 예제에서 Ironman에 있는 jarvis를 Mark1, Mark4가 override합니다.
* Mark42에는 jarvis가 없으므로, Ironman의 내용을 그대로 사용하게 됩니다.
* for반복문에서 in안에 3개의 객체가 들어있는데, marks의 객체형이 동적으로 결정됩니다. 따라서 그때마다 jarvis가 각각의 객체에 맞게 다르게 사용됩니다.

#[3번 문제]
다음 각 요구사항 모두를 만족시키는 Counter 클래스를 코딩하시오 (정답을 각 요구사항별로 입력할 필요 없이 3번 문제에 대해 1개의 클래스 정의 코드를 제시하면 된다.)

요구사항 1. 생성자에 count 값과 step 값을 인자로 받을 수 있다.

    count: Counter 인스턴스가 지니는 초기 정수 값
    step: Counter 인스턴스의 count가 증가되는 증분 (defalt 값: 1)
        >>> c = Counter(10) 
        >>> d = Counter(10, 2) 

    => 위 요구사항에 대해 클래스 내에 __init__(self, cnt, stp=1)를 정의하였습니다.
       값을 인자로 받아와(cnt, stp) 생성자에 count값과 step값으로 정하였습니다. step의 defalt는 1 이므로 stp=1을 하였습니다.

요구사항 2. 다음과 같이 Counter의 인스턴스를 출력을 해주는 __str__() 메소드를 Counter 클래스 내에 구현하시오.

        >>> print c 
        [Count (step: 1)] 10 
        >>> print d 
        [Count (step: 2)] 10 

    => 위 요구사항에 대해 클래스 내에 __str__(self)을 정의하였습니다. 
    return '[Count (step : %s)] %s' % (self.step, self.count)를 함으로써 주어진 양식에 맞게 Counter의 인스턴스를 출력하도록 하였습니다. Counter의 인스턴스 이므로 self를 적어주었습니다. 

요구사항 3. 다음과 같이 step에 주어진 증분만큼 count를 증가시키는 incr() 메소드를 Counter 클래스 내에 구현하시오.

        >>> c.incr() 
        >>> d.incr() 
        >>> print c 
        [Count (step: 1)] 11 
        >>> print d 
        [Count (step: 2)] 12

    => 위 요구사항에 대해 클래스 내에 incr(self)함수를 정의하였습니다.
    위 함수는 Counter의 인스턴스의 step에 주어진 증분만큼 count를 증가시켜야 하므로 self.count = self.count + self.step 입니다. 

요구사항 4. Counter 클래스 내에 관련 메소드를 추가하여 인스턴스 객체를 직접 호출(call)할 수 있도록 하시오. 인스턴스 객체를 직접 호출했을 때에 내부적으로 incr() 메소드를 호출하는 방법으로 구현하시오.

        >>> c() 
        >>> d() 
        >>> print c 
        [Count (step: 1)] 12 
        >>> print d 
        [Count (step: 2)] 14 

    => 위 요구사항에 대해 클래스 내에 __call__(self)를 정의하였습니다.
    그 안에 self.incr()를 넣음으로써, 인스턴스 객체를 직접 호출했을때, incr()메소드를 호출하도록 하였습니다. 인스턴스 객체 내의 일 이므로 self를 사용하였습니다.

요구사항 5. 다음과 같은 두 개의 산술 연산 (+, -)이 수행될 수 있도록 Counter 클래스 내에 관련 메소드를 추가하시오.

        >>> c = c + 5 
        >>> d = d - 5 
        >>> print c 
        [Count (step: 1)] 17 
        >>> print d 
        [Count (step: 2)] 9 

    => 위 요구사항에 대해 클래스 내에 __add__(self, other)와 def __sub__(self, other)를 정의하였습니다.
    각 내부는 return Counter(self.count + other, self.step), return Counter(self.count - other, self.step)와 같은데, +연산을 하였을 경우, 인스턴스(self)의 count값과 연산되는 값을 더한것(혹은 뺀 것)을 새로 만들어지는 인스턴스의 count값으로, 그리고 기존의 step을 이용하면 제시된 두 개의 산술 연산이 수행됩니다.

요구사항 6. 다음과 같은 관계연산 (>, <, ==)이 수행될 수 있도록 Counter 클래스 내에 관련 메소드를 추가하시오.

    >>> print c > 10 
    True 
    >>> print d > 10 
    False 
    >>> print c < 10 
    False 
    >>> print d < 10 
    True 
    >>> print c == 17 
    True 
    >>> print d != 9 
    False 

    => 관계 연산을 위해 __gt__(self, other), __lt__(self, other), __eq__(self, other), __ne__(self, other)를 정의해주었습니다.
    함수 내용은 return self.count > other과 같은데, T/F가 출력되므로 반환값은 비교한 것을 반환해주면 됩니다. 각 연산에 맞게 >, <, ==, !=를 넣어줌으로써 수행될 수 있도록 하였습니다.

In [16]:
class Counter:
    def __init__(self, cnt, stp=1):
        self.count = cnt
        self.step = stp
        
    def __str__(self):
        return '[Count (step : %s)] %s' % (self.step, self.count)
    
    def incr(self):
        self.count = self.count + self.step
        
    def __call__(self):
        self.incr()
        
    def __add__(self, other):
        return Counter(self.count + other, self.step)
    def __sub__(self, other):
        return Counter(self.count - other, self.step)
    
    def __gt__(self, other):
        return self.count > other
    def __lt__(self, other):
        return self.count < other
    def __eq__(self, other):
        return self.count == other
    def __ne__(self, other):
        return self.count != other
        
c = Counter(10)
d = Counter(10, 2)

print c
print d
print
        
c.incr()
d.incr()
print c
print d
print

c()
d()
print c
print d
print

c = c + 5 
d = d - 5 
print c
print d
print

print c > 10
print d > 10
print c < 10
print d < 10
print c == 17
print d != 9

[Count (step : 1)] 10
[Count (step : 2)] 10

[Count (step : 1)] 11
[Count (step : 2)] 12

[Count (step : 1)] 12
[Count (step : 2)] 14

[Count (step : 1)] 17
[Count (step : 2)] 9

True
False
False
True
True
False


#[4번 문제]
다음은 내장 자료형 list를 서브클래싱하여 만든 MySet 클래스 정의 내용이다. 다음 클래스 정의에서 __init__(), __str()__(), elimicate_duplicate()의 세 개의 메소드 코드 내용을 자신이 다른 사람에게 가르친다고 생각하며 설명해보시오.

MySet은 집합(Set) 자료형을 정의하려는 의도하에 만들어진 클래스이다.

In [17]:
class MySet(list):
    def __init__(self, l):
        for e in l:
            self.append(e)
        MySet.eliminate_duplicate(self)
    
    def __str__(self):
        result = "MySet: {"
        for e in self:
            result = result + str(e) + " ,"
        result = result[0:len(result)-2] + "}"
        return result

    @staticmethod    
    def eliminate_duplicate(l):
        s = []
        for e in l:
            if e not in s:
                s.append(e)
        l[:] = []
        for e in s:
            l.append(e)                     
    
if __name__ == "__main__":
    s = MySet([1, 2, 2, 3])
    print s
    t = MySet([2, 3, 4, 5, 6, 7, 8, 8, 8, 8, 8, 9])
    print t

MySet: {1 ,2 ,3}
MySet: {2 ,3 ,4 ,5 ,6 ,7 ,8 ,9}


* eliminate_duplicate는 정적 메소드로 인스턴스 객체와 무관하게 클래스 이름 공간에 존재하는 메소드로서 클래스 이름을 이용하여 직접 호출할 수 있는 메소드입니다. 장식자 @staticmethod 를 통해 정적 메소드임을 알 수 있습니다. 위 메소드는 인자로 받은 리스트 l를 그 안에서 중복된 원소를 뺀 리스트로 다시 만들어 주는 역할을 하는 메소드입니다. 처음 반복문에서 l에서 중복되지 않은 원소들을 e로 뽑아내고, l을 비운 다음, e를 다시 l에 넣는 두번째 반복문을 볼 수 있습니다.
* __init__(self, l)를 통해 인자로 들어온 l의 원소를 갖는 인스턴스를 append()함수를 이용하여 만들고, eliminate_duplicate()함수를 이용하여 중복된 원소를 제거해줍니다.
* __str__(self)를 통해 출력시 적절한 모양(?)으로 만들어 줍니다. 반복문을 이용해서 result에 '원소값, '를 더해주고, 반복문이 끝난 후 마지막 원소 뒤의 ' ,'를 빼고 '}'를 넣음으로써 적절한 형태가 완성됩니다. 따라서 출력할 경우, MySet: {1 ,2 ,3} 과 같은 형태로 출력됩니다.


#[5번 문제]
4번 문제에 정의된 MySet 클래스에 메소드를 추가하여 다음 각 요구사항 모두를 만족시키는 코딩을 제시하시오

정답을 각 요구사항별로 입력할 필요 없이 요구사항 3개 전체에 대해 1개의 MySet 클래스 정의 코드를 제시하고 주석등으로 해당 코드를 설명한다.

요구사항 1. | 연산으로 두 집합의 합집합을 반환한다.

    >>> u = s | t 
    >>> print u 
    MySet: {1, 2, 3, 4, 5, 6, 7, 8, 9} 

    => 클래스 내에 __or__(self, other)를 정의하여, 요구사항을 수행합니다. 위의 예에서 self가 s, other가 t라고 생각하면 됩니다. __init__와 유사한 방법으로, 우선 temp에 self와 other의 원소를 추가한 후 eliminate_duplicate(self)를 이용하여 중복된 원소를 제거하면 합집합이 되며, 그 합집합을 반환하면 됩니다.

요구사항 2. & 연산으로 두 집합의 교집합을 반환한다.

    >>> u = s & t 
    >>> print u 
    MySet: {2, 3} 

    => 클래스 내에 __and__(setf, other)를 정의하여 요구사항을 수행합니다. sumset이라는 함수를 따로 만들어서 두 인스턴스의 원소중 겹치는 것을 뽑아내는 과정을 수행합니다. set1을 수행한 결과값을반환하면 됩니다. set1은 eliminate_duplicate와 같은 정적 메소드로, self의 원소를 반복하는데, other에 들어있는 원소 중 같은것이 있다면 s에 추가합니다. l(= self)를 비운 후, s의 원소를 저장하고 반환해줍니다.

요구사항 3. - 연산으로 두 집합의 차집합을 반환한다.

    >>> s = MySet([1, 2, 3]) 
    >>> t = MySet([3, 4, 5]) 
    >>> u = s - t 
    >>> print u 
    MySet: {1} 

####=> 차집합인 s-t일 경우, s의 원소 중 s와 t의 교집합의 원소에 들어가는 것을 뺀 나머지라 {1, 2}가 나와야 한다고 생각하기에 그런 방식으로 풀었습니다. 

=> 클래스 내에 __sub__(self, other)를 정의하여 요구사항을 수행합니다. 우선 temp1에는 self의 원소를 넣고, temp2에는 앞에 만들었던 set1함수를 이용하여 교집합을 저장합니다. 그 다음 set2함수의 결과를 반환합니다. set2함수는 정적 메소드로, 빈 리스트 s에 만약 temp1의 원소가 temp2의 원소에 없다면 s에 추가합니다. s는 차집합이고 그것을 MySet2(s)로 반환하면 원하는 답이 반환됩니다. 

In [18]:
class MySet2(list):
    def __init__(self, l):
        for e in l:
            self.append(e)
        MySet2.eliminate_duplicate(self)
    
    def __str__(self):
        result = "MySet2: {"
        for e in self:
            result = result + str(e) + " ,"
        result = result[0:len(result)-2] + "}"
        return result
    
    def __or__(self, other):
        temp = []
        for e in self:
            temp.append(e)        
        for h in other:
            temp.append(h)
        MySet2.eliminate_duplicate(temp)
        return MySet2(temp)
    
    def __and__(self, other):
        return MySet2.set1(self, other)
    
    def __sub__(self, other):
        temp = []
        for e in self:
            temp.append(e)  
        temp2 = MySet2.set1(self, other)
        return MySet2.set2(temp, temp2)
        
        

    @staticmethod    
    def eliminate_duplicate(l):
        s = []
        for e in l:
            if e not in s:
                s.append(e)
        l[:] = []
        for e in s:
            l.append(e)  
            
    @staticmethod    
    def set1(l, k):
        s = []
        for e in l:
            if e in k:
                s.append(e)
        l[:] = []
        for e in s:
            l.append(e) 
        return l
            
    @staticmethod    
    def set2(t1, t2):
        s = []
        for e in t1:
            if e not in t2:
                s.append(e)
        return MySet2(s)
    
if __name__ == "__main__":
    s = MySet2([1, 2, 2, 3])
    print s
    t = MySet2([2, 3, 4, 5, 6, 7, 8, 8, 8, 8, 8, 9])
    print t
    u = s | t
    print u 
    u = s & t
    print u 
    
    s = MySet2([1, 2, 3])
    t = MySet2([3, 4, 5]) 
    u = s - t
    print u 

MySet2: {1 ,2 ,3}
MySet2: {2 ,3 ,4 ,5 ,6 ,7 ,8 ,9}
MySet2: {1 ,2 ,3 ,4 ,5 ,6 ,7 ,8 ,9}
MySet2: {2 ,3}
MySet2: {1 ,2}


#[6번 문제]
5번 문제에서 정의한 MySet 클래스에 대해 다음 예제를 수행하면 오류없이 올바르게 동작하는 것을 확인할 수 있다. 다음 예제 내에 있는 len(), bool() 내장함수와 in 키워드 사용 예제가 별다른 메소드 정의를 하지 않았는 데도 올바르게 수행되는 이유를 설명하시오.

> \>\>\> s = MySet([1, 2, 3, 4, 5, 6])  <br/>
> \>\>\> print len(s)  <br/>
> 6 <br/>
> \>\>\> print bool(s)  <br/>
> True <br/>
> \>\>\> print 2 in s  <br/>
> True <br/>

In [29]:
s = MySet([1, 2, 3, 4, 5, 6]) 
print len(s) 

t = MySet([11, 12, 13, 14, 15, 16]) 
print len(t)

print s + t
print bool(s)
print 2 in s

print dir()
print dir(list)
print dir(__builtin__)

6
6
[1, 2, 3, 4, 5, 6, 11, 12, 13, 14, 15, 16]
True
True
['Counter', 'In', 'Ironman', 'M', 'Mark1', 'Mark42', 'Mark7', 'MySet', 'MySet2', 'OrderedWebWordsFrequency', 'OrderedWebWordsFrequency2', 'Out', 'WebWordsFrequency', '_', '__', '___', '__builtin__', '__builtins__', '__doc__', '__name__', '__package__', '_dh', '_i', '_i1', '_i10', '_i11', '_i12', '_i13', '_i14', '_i15', '_i16', '_i17', '_i18', '_i19', '_i2', '_i20', '_i21', '_i22', '_i23', '_i24', '_i25', '_i26', '_i27', '_i28', '_i29', '_i3', '_i4', '_i5', '_i6', '_i7', '_i8', '_i9', '_ih', '_ii', '_iii', '_oh', '_sh', 'a', 'b', 'c', 'c1', 'c2', 'c3', 'c4', 'cnt', 'col', 'd', 'days', 'exit', 'f1', 'fac', 'get_ipython', 'i', 'line', 'marks', 'n2', 'num', 'nums', 'operator', 'quit', 'res', 'result', 'row', 's', 's1', 's2', 's3', 'startD', 'string', 'sumdays', 't', 'temp', 'u', 'urllib2', 'vals', 'w1', 'w2', 'w3', 'w4', 'w5', 'year']
['__add__', '__class__', '__contains__', '__delattr__', '__delitem__', '__delslice__', '__doc__', '_

=> 간단히 말하면 미리 정의되어있기 때문에 별다른 메소드를 정의하지 않아도 가능한 것 입니다. dir()을 하면 현재이름공간에 존재하는 이름 리스트를 반환하는데, 그안에 보면 __builtin__ 안에 len과 bool이 들어있는 것을 확인할 수 있습니다. MySet은 리스트를 상속받는 클래스로, 리스트의 내장 함수를 모두 사용할 수 있기 때문입니다. in 또한 예약어로, 변수명으로 사용할 수 없고 미리 정의되어있는 단어로, 따로 정의하지 않아도 사용가능한 것 입니다.

#[7번 문제]
###프로젝트오일러 16번
2^15 = 32768 의 각 자리수를 더하면 3 + 2 + 7 + 6 + 8 = 26 입니다.

2^1000의 각 자리수를 모두 더하면 얼마입니까?

In [21]:
result = 0
for s in str(2**1000):
    result += int(s)
print result

1366


=> 2^1000을 문자열로 바꿔서, 각 자리수를 원소로 취급하여 반복문을 통해 반복합니다. 반복문이 반복될 때마다 각 자리수 하나를 만나게 되는데 그것을 int()를 이용하여 문자가 아닌 정수로 계산하여 result에 누적으로 더한다면, 문제가 원하는 답이 나옵니다. 

#[8번 문제]
### 프로젝트오일러 17번
1부터 5까지의 숫자를 영어로 쓰면 one, two, three, four, five 이고,
각 단어의 길이를 더하면 3 + 3 + 5 + 4 + 4 = 19 이므로 사용된 글자는 모두 19개입니다.

1부터 1,000까지 영어로 썼을 때는 모두 몇 개의 글자를 사용해야 할까요?

참고: 빈 칸이나 하이픈('-')은 셈에서 제외하며, 단어 사이의 and 는 셈에 넣습니다.

      예를 들어 342를 영어로 쓰면 three hundred and forty-two 가 되어서 23 글자,
      115 = one hundred and fifteen 의 경우에는 20 글자가 됩니다.

In [22]:
c1 = [3, 3, 5, 4, 4, 3, 5, 5, 4] 
c2 = [3, 6, 6, 8, 8, 7, 7, 9, 8, 8] 
c3 = [6, 6, 5, 5, 5, 7, 6, 6] 
c4 = [7, 3, 8]

# 1~99
s1 = sum(c1)*9 + sum(c2) + sum(c3)*10

# 101~199, ..., 901~999
s2 = s1*9 + sum(c1)*99 + sum(c4[:2])*99*9

# 100, 200, ..., 1000
s3 = sum(c1) +  c4[0]*9 + c1[0] + c4[2]
print s1 + s2 + s3

21124



=> 우선 1~9, 10~19, 20~90, 100 and 1000 그룹으로 영어의 글자 수를 각각 c1, c10, c20, c100에 넣는다. s1은 1~99까지의 글자 수로, c1(1~9)이 총 9번 반복되고(1~9, 21~29, 31~39, 41~49, 51~59, 61~69, 71~79, 81~89, 91~99), c10은 10, 11, .., 19까지의 특이한 경우이므로 1번 반복되고, 이후의 10단위인 c20(20, 30, 40, ...)은 각 자리수마다 10번씩 반복된다. 따라서 s1 = sum(c1)9 + sum(c10) + sum(c20)10 이 된다. s2는 101~199, 201~299, ... 901, 999의 그룹으로, (1)one / (2)hundred and / (3)one처럼 나누어 보면, 크게는 s1그룹에서 hundred and가 추가된 것과 같이 생각하면되는데, (3)은 1~99가 9번 반복되고, (2)는 100단위당 hundred and가 99번 반복되고(101, 102, ..., 199) 그것이 100~900까지 9번 반복된다. (1)은 1~9가 99번 반복되므로 s2 = s19 + sum(c1)99 + sum(c100[:2])999가 된다. s3은 100, 200,...,1000으로, one / hundred 처럼 1~9가 1번 반복되고, hundred가 9번 반복되며 1000은 one thousand 이므로 그것을 더하면 s3 = sum(c1) + c100[0]*9 + c1[0] + c100[2]이 된다. 결과는 s1 + s2 + s3 = 21124이다.

#[9번 문제]
###프로젝트오일러 18번
다음과 같이 삼각형 모양으로 숫자를 배열했습니다.

3
7 4
2 4 6
8 5 9 3

삼각형의 꼭대기부터 아래쪽으로 인접한 숫자를 찾아 내려가면서 합을 구하면, 위의 그림처럼 3 + 7 + 4 + 9 = 23 이 가장 큰 합을 갖는 경로가 됩니다.

다음 삼각형에서 합이 최대가 되는 경로를 찾아서 그 합을 구하세요.

75
95 64
17 47 82
18 35 87 10
20 04 82 47 65
19 01 23 75 03 34
88 02 77 73 07 63 67
99 65 04 28 06 16 70 92
41 41 26 56 83 40 80 70 33
41 48 72 33 47 32 37 16 94 29
53 71 44 65 25 43 91 52 97 51 14
70 11 33 28 77 73 17 78 39 68 17 57
91 71 52 38 17 14 91 43 58 50 27 29 48
63 66 04 68 89 53 67 30 73 16 69 87 40 31
04 62 98 27 23 09 70 98 73 93 38 53 60 04 23

참고: 여기서는 경로가 16384개밖에 안되기 때문에, 모든 경로의 합을 일일이 계산해서 답을 구하는 것이 가능합니다.
하지만 67번 문제에는 100층짜리 삼각형 배열이 나옵니다. 그런 경우에는 좀 더 현명한 풀이 방법을 찾아야겠지요.

In [23]:
nums = '''
75
95 64
17 47 82
18 35 87 10
20 04 82 47 65
19 01 23 75 03 34
88 02 77 73 07 63 67
99 65 04 28 06 16 70 92
41 41 26 56 83 40 80 70 33
41 48 72 33 47 32 37 16 94 29
53 71 44 65 25 43 91 52 97 51 14
70 11 33 28 77 73 17 78 39 68 17 57
91 71 52 38 17 14 91 43 58 50 27 29 48
63 66 04 68 89 53 67 30 73 16 69 87 40 31
04 62 98 27 23 09 70 98 73 93 38 53 60 04 23
'''

#리스트로 나누기
vals = []
for line in nums.strip('\n').split('\n'):
    temp = []
    for num in line.split():
        temp.append(int(num))
    vals.append(temp)
    
#더하기
for row in range(13, -1, -1):
    for col in range(row+1):
        vals[row][col] += max(vals[row+1][col], vals[row+1][col+1])
        
print vals[0][0]

1074


=> 처음에는 일일이 다 더하는 방법으로 했었는데, 맨아랫줄의 바로 윗줄 부터 자신의 아래층 중 좌/우 값 중 큰 값을 더하면서 맨 위까지 올라가면, 맨 위 값이 가장 큰 값이라는 점을 보고 문제를 풀었습니다. 우선, 문자열로 받았기 때문에 각 값들을 리스트로 만들어주었습니다. 그 다음, 마지막에서 바로 윗줄이므로 인덱스는 13부터, 위로 올라가므로 -1씩, 제일 윗줄까지 해야하므로 -1까지 하는데, 아래층 중 좌/우 값을 비교하여 큰 값을 더해야 하므로 max(vals[row+1][col], vals[row+1][col+1])가 되고, 그것을 vals[row][col]에 더해줍니다. 이렇게 반복하여 제일 윗줄까지 가면 맨 위의 값이 가장 큰 값이 됩니다.

#[10번 문제]
###프로젝트오일러 19번
다음은 달력에 관한 몇 가지 일반적인 정보입니다 (필요한 경우 좀 더 연구를 해 보셔도 좋습니다).

1900년 1월 1일은 월요일이다.
4월, 6월, 9월, 11월은 30일까지 있고, 1월, 3월, 5월, 7월, 8월, 10월, 12월은 31일까지 있다.
2월은 28일이지만, 윤년에는 29일까지 있다.
윤년은 연도를 4로 나누어 떨어지는 해를 말한다. 하지만 400으로 나누어 떨어지지 않는 매 100년째는 윤년이 아니며, 400으로 나누어 떨어지면 윤년이다
20세기 (1901년 1월 1일 ~ 2000년 12월 31일) 에서, 매월 1일이 일요일인 경우는 총 몇 번입니까?

In [24]:
#일요일 = 0
M = [31,28,31,30,31,30,31,31,30,31,30,31]
startD = 2 #화요일
cnt=0

for year in range(1901,2001) :
    #윤년 계산
    if(year % 4 == 0):
        if ((year % 100 == 0) and (year % 400 != 0)) :
            M[1]=29
    else :
        M[1]=28

    #월별 요일 계산
    for days in M :
        sumdays = startD + days
        startD = sumdays % 7
        if(startD == 0):
            cnt += 1

print cnt

171


=> 요일을 일요일은 0, 월요일은 1 ... 로 정한다면, 예를들어 월의 첫일이 수요일이고 그 월이 30일까지 있다고 한다면, 그 다음달의 첫째날 요일은 (3+30)/7을 한 결과의 몫이 될 것입니다. 
우선, 1901년부터 2000년까지를 구해야 하므로 for year in range(1901,2001)이며, 문제에 주어진 윤년의 조건에 맞춰서 윤년이라면 2월의 날 수를 29로 바꾸고, 아니라면 28로 합니다. 그 다음, 요일을 계산해야하는데, 우선 이 달의 시작 요일과 이 달의 날 수를 더하고, 거기서 7로 나눈 다음 그 몫을 시작 요일로 합니다. 이것은 다음 달의 1일의 요일이 되는데, 만약 0이라면 1일이 일요일이라는 말이므로, cnt 수를 1 증가시킵니다.
모든 반복문이 끝나고 최종 cnt가 문제의 답이 됩니다.

#[11번 문제]
###프로젝트오일러 20번
n! 이라는 표기법은 n × (n − 1) × ... × 3 × 2 × 1을 뜻합니다.

예를 들자면 10! = 10 × 9 × ... × 3 × 2 × 1 = 3628800 이 되는데,
여기서 10!의 각 자리수를 더해 보면 3 + 6 + 2 + 8 + 8 + 0 + 0 = 27 입니다.

100! 의 자리수를 모두 더하면 얼마입니까?

In [25]:
#팩토리얼 함수
def fac(k):
    n = 1
    for i in range(1, k + 1):
        n *= i
    return n

res = 0
num = str(fac(100))

#자리수 더하기
for i in num:
    res += int(i)

print res

648


=> 팩토리얼 함수를 정의해 놓은 후(이전 과제 참고), 100!을 구한 값을 문자열로 받고, 그 문자열에서 원소들을 반복하는데, 원소는 각 자리수가 됩니다. 따라서 그 값들을 숫자로 받아와 더하면 문제의 정답이 됩니다.

#[12번 문제]
###프로젝트오일러 21번
n의 약수들 중에서 자신을 제외한 것의 합을 d(n)으로 정의했을 때,
서로 다른 두 정수 a, b에 대하여 d(a) = b 이고 d(b) = a 이면 
a, b는 친화쌍이라 하고 a와 b를 각각 친화수(우애수)라고 합니다.

예를 들어 220의 약수는 자신을 제외하면 1, 2, 4, 5, 10, 11, 20, 22, 44, 55, 110 이므로 그 합은 d(220) = 284 입니다.
또 284의 약수는 자신을 제외하면 1, 2, 4, 71, 142 이므로 d(284) = 220 입니다.
따라서 220과 284는 친화쌍이 됩니다.

10000 이하의 친화수들을 모두 찾아서 그 합을 구하세요.

In [26]:
#약수 합 구하기
def f1(a):
    n = 1
    temp = []
    while(n < a):
        if(a % n == 0):
            temp.append(n)
        n += 1
    return sum(temp)

result = []

#친화수의 합 구하기
for a in range(2, 10001):
    b = f1(a)
    n2 = f1(b)
    if(a == n2 and a != b):
        result.append(a)
        result.append(b)
result = list(set(result))
print result
print sum(result)

[1184, 6368, 220, 5020, 2924, 6232, 1210, 5564, 284, 2620]
31626


=> 우선 f1은 자신을 제외한 약수의 합을 구하는 함수입니다. 1부터 입력된 수보다 작을때 까지 반복하는데, 나누었을때 0이라면 temp 리스트에 추가하고 , 반복문이 끝났을 때, temp리스트의 합입니다.
그 다음, 2부터 10000까지 반복하는데, f1(a)의 값이 a의 약수중 자신을 제외한 약수의 합이고, 그 합을 다시 f1(b)했을때 나오는 값이 a와 같다면 a와 b는 친화쌍일 것 입니다. 단, a와b가 같은 수일 경우가 있으므로 그 부분을 조건문을 이용하여 제외합니다. 
중복된 수가 들어갈 가능성이 있으므로 set()과 list()를 이용하여 제거해주고 그 리스트의 합이 문제의 정답입니다.

#[13번 문제]
이전 Assignment 3 (Assignment 4가 아님)의 마지막 문제는 웹 URL로 지정된 웹페이지를 문자열로 가져와 모든 HTML 태그 및 CSS와 Javascript를 제외한 순수 텍스트를 얻어내고 그 안에 존재하는 단어를 추출하여 각 단어들에 대해 출현빈도를 사전형태({'world': 2, 'hello': 1, 'python': 1})로 저장하여 출력하는 것이었다. 이번에는 Assignment 3을 다시 확장/변형하여 다음과 같은 조건을 만족하도록 구현하시오.

    1) 새로운 클래스 WebWordsFrequency를 정의하시오.
        이전 Assignment 4에서 고려했던 pickle 모듈등은 전혀 고려하지 마시오.
        즉, 하드시스크에 저장하고 다시 로드하는 기능은 본 숙제에는 배제하시오.
        Assignment 4에서 고려했던 영어 불용어 처리는 적용하시오.
        영어 불용어: [ 'a', 'and', 'are', 'as', 'at', 'be', 'but', 'by', 'for', 'if', 'in', 'into', 'is', 'it', 'no', 'not', 'of', 'on', 'or', 's', 'such', 't', 'that', 'the', 'their', 'then', 'there', 'these', 'they', 'this', 'to', 'was', 'will', 'with']
        Assignment 4에서 처럼 국문 사이트가 아닌 영어 사이트만을 고려하여 아래 사항을 코딩하시오.

2) 생성자에 URL을 0개에서 임의의 개수를 넣을 수 있도록 생성자 인수를 가변인수로 정의하여 각각의 URL을 리스트 자료형에 유지하시오.
    >>> w1 = WebWordsFrequency('http://www.cnn.com', 'http://www.times.com', 'https://www.amazon.com') 
    >>> w2 = WebWordsFrequency('http://www.cnn.com', 'http://www.times.com') 
    >>> w3 = WebWordsFrequency() 

3) addUrl() 메소드를 구현하여 인스턴스를 생성한 이후에도 URL을 추가할 수 있도록 한다.

    반드시 1개의 URL을 추가하도록 구현 (즉, 동시에 여러 개의 URL을 추가하는 것은 배제)
        >>> w1.addUrl('https://github.com')
        >>> w3.addUrl('http://stackoverflow.com')

4) removeUrl() 메소드를 구현하여 URL을 삭제할 수 있도록 한다.

    반드시 1개의 URL을 삭제하도록 구현 (즉, 동시에 여러 개의 URL을 삭제하는 것은 배제)
        >>> w1.removeUrl('http://www.cnn.com')
        >>> w2.removeUrl('http://stackoverflow.com')

5) listUrls() 메소드를 구현하여 현재 등록된 모든 URL을 출력하는 기능을 추가
    >>> w1.listUrls()
        http://www.times.com 
        https://www.amazon.com 
        https://github.com 

6) getWordsFrequency() 메소드를 구현하여 각 URL의 웹페이지들을 종합적으로 분석한 단어 출현 빈도 사전을 반환하시오.

    만약 등록된 URL이 없다면 공백 사전을 반환
        >>> w1.getWordsFrequency() 
        {'hello': 8, 'site': 12, 'world': 2, 'science': 11, 'program': 1, 'python': 1}

7) getMaxFreqencyWords() 메소드를 구현하여 각 URL의 웹페이지들을 종합적으로 분석한 단어 출현 빈도 사전에서 가장 많이 출현한 단어 리스트를 반환하시오.

    >>> w1.getMaxFreqencyWords() 
        site 
    최다 출현 단어의 빈도수가 동일한 경우 모두 출력해주어야 함
    >>> w2.getMaxFreqencyWords() 
        site 
        science
    만약 등록된 URL이 없다면 None을 반환
    
#####=> 위 문제에서는 문제에서 리스트를 반환하라 하여서, 출력시 예시처럼이 아닌 리스트 형식으로 출력되도록 하였습니다.

=> 요구사항 별로 소스 설명을 작성하기가 번거로워서 코드에 주석으로 설명을 대체하였습니다.

In [27]:
import urllib2
import string
import operator

class WebWordsFrequency:
    # 요구사항2번) 생성자에 가변 인수 설정
    def __init__(self, *s):
        self.urllist = list(s) # url 리스트
        self.cnt = len(list(s)) # url 보유 숫자
        
    #요구사항7번) 가장 많이 출현한 단어 리스트 반환    
    def getMaxFreqencyWords(self):
        self.getWordsFrequency() # 단어 사전을 위해 호출
        if self.cnt == 0: # 등록된 URL없는경우
            return None
        else :
            self.resultlist = []
            self.sorted_dic = sorted(self.dic.iteritems(), key=operator.itemgetter(1)) #사전 크기순으로 정렬
            for q in range(1, self.cnt):
                if(self.sorted_dic[-1][1] == self.sorted_dic[-q][1]): #숫자 같은지 확인
                    self.resultlist.append(self.sorted_dic[-q][0])  #같을 경우 결과 리스트에 추가
                else:
                    break
            return self.resultlist 
    
    #요구사항6번) 단어 출현 빈도 사전 - 이전 과제 참고
    def getWordsFrequency(self):
        self.dic = dict() #사전 생성
        for k in range(self.cnt):
            source = urllib2.urlopen(self.urllist[k]).read()

            for i in range(0, source.count('<!--')):
                if source.find("<!--") != -1:
                    source = source.replace(source[source.find("<!--"):source.find("-->")+3:],"")

            for i in range(0,source.count("<")):
                if source.find("<") != -1:
                    source = source.replace(source[source.find("<"):source.find(">")+1:],"")
                    
            words = source.split()
            s = ' '.join(words)
            
            #구두문자 제거
            words = list(s)
            for i in range(len(words)):    
                for k in range(len(string.punctuation)):
                    if(words[i] == string.punctuation[k]):
                        words[i] = ' '
            words = ''.join(words)

            #구두문자가 제거된 문자열을 공백을 제외한 리스트로 만들기
            words = words.split()
            
            #영어 불용어 제거
            eng = ['a', 'and', 'are', 'as', 'at', 'be', 'but', 'by', 'for', 'if', 'in', 'into', 'is', 'it', 'no', 'not', 'of', 'on', 'or', 's', 'such', 't', 'that', 'the', 'their', 'then', 'there', 'these', 'they', 'this', 'to', 'was', 'will', 'with']
            for i in range(len(words)):
                for k in range(len(eng)):
                    if words[i] == eng[k]:
                        words[i] = ' '
            words = ' '.join(words)

            #영어 불용어가 제거된 문자열을 공백을 제외한 리스트로 만들기
            words = words.split()
            
            #사전 만들고 카운트 세기
            for word in words:
                if word not in self.dic:
                    self.dic[word] = 1
                else:
                    self.dic[word] += 1                  
            return self.dic        
        
    # 요구사항3번) URL을 추가
    def addUrl(self, au):
        self.urllist.append(au) #url 리스트에 추가
        self.cnt += 1 #url 보유 숫자 증가

    # 요구사항4번) URL을 삭제
    def removeUrl(self, ru):
        check = 0
        #삭제할 url이 있는지 확인
        for m in range(self.cnt):
            if(self.urllist[m] == ru):
                check = 1
                break
            
        #삭제하고자 하는 url이 등록되어있지 않는 경우
        if (check == 0):
            print "Not found"
            return
        
        #있는 경우
        ind = self.urllist.index(ru) #인덱스값 찾기
        self.urllist.pop(ind) #제거
        self.cnt -= 1 #url 보유 숫자 감소
      
    # 요구사항5번) URL을 출력
    def listUrls(self):
        for k in range(self.cnt):
            print self.urllist[k]
            
        # 보유한 url이 없는 경우
        if (self.cnt == 0):
            print "No Url"
        
w1 = WebWordsFrequency('http://www.cnn.com', 'http://www.times.com', 'https://www.amazon.com') 
w2 = WebWordsFrequency('http://www.cnn.com', 'http://www.times.com') 
w3 = WebWordsFrequency()

w1.addUrl('https://github.com')
w3.addUrl('http://stackoverflow.com')

w1.removeUrl('http://www.cnn.com')
w3.removeUrl('http://stackoverflow.com')
w3.removeUrl('http://www.cnn.com')
print
w1.listUrls()
print
w2.listUrls()
print
w3.listUrls()
print
print
print w1.getWordsFrequency() 
print
print
print w1.getMaxFreqencyWords()
print w2.getMaxFreqencyWords() 
print w3.getMaxFreqencyWords()

Not found

http://www.times.com
https://www.amazon.com
https://github.com

http://www.cnn.com
http://www.times.com

No Url


{'todays': 2, 'hpmiddle': 1, '6px': 3, 'NYT': 6, 'Dining': 1, 'Does': 2, 'asian': 1, 'aides': 1, 'domContent': 1, 'Appeal': 1, 'whose': 1, 'queueTime': 1, 'paris': 4, '09736473': 1, 'sourceApp': 1, 'invaders': 1, 'Vengeance': 2, 'China': 1, 'Cases': 1, 'Workers': 2, '100000003980215': 1, 'nyregion': 15, '100000004068048': 1, 'void': 1, 'Inv3': 1, 'Inv2': 1, 'Inv1': 1, 'chicanery': 1, 'Military': 4, 'Streetwear': 2, 'vast': 1, 'alpine': 2, 'school': 3, 'snowflakes': 1, '100000004072430': 1, 'ref': 3, '4738': 1, 'cmd': 2, '100000004072438': 3, 'Feel': 1, 'Go': 4, 'Paul': 2, 'heading': 199, '291': 1, 'bill': 3, 'force': 3, 'POWELL': 1, '0052': 1, '0051': 1, 'weddings': 1, 'Negative': 1, '9C34': 1, 'second': 1, 'street': 2, 'Theater': 4, 'n': 64, 'alt': 49, '0202D0E4': 1, 'C59B': 1, 'Lamar': 2, 'noticeError': 1, 'Creating': 1, 'Merkel': 1, 'selected': 1, 'asia': 4, '

#[14번 문제]
13번 문제에서 정의한 WebWordsFrequency 클래스를 상속하여 OrderedWebWordsFrequency 클래스를 정의하고 슈퍼클래스에 정의된 getWordsFrequency() 메소드를 오버라이드 하여 단어 출현 빈도를 내림 차순으로 정렬하여 리스트로 출력하시오.

    리스트 내의 각 원소는 단어와 빈도를 쌍으로 지니는 튜플이다.
    getWordFrequence() 메소드에 reverse 라는 인자를 만들고 true 또는 false를 인자로 받을 수 있도록 한다.

    reverse 인자의 디폴트 값은 false 이며, 기본적으로 내림차순으로 정렬한다.
    reverse 인자에 true를 넣으면 오름차순으로 정렬한다.
        >>> w4 = OrderedWebWordsFrequency('http://www.times.com', 'https://www.amazon.com', 'https://github.com')

        >>> w4.getWordsFrequency()
            [('site', 12), ('science', 11), ('hello', 8), ('world', 2), ('program', 1), ('python', 1)]

        >>> w4.getWordsFrequency(reverse=true)
            [('program', 1), ('python', 1), ('world', 2), ('hello', 8), ('science', 11), ('site', 12)]

####=> 문제에서 true, false를 소문자로 제시하셨는데, True, False를 나타내는 것이라 생각하여 대문자로 문제를 풀었습니다.

In [28]:
class OrderedWebWordsFrequency(WebWordsFrequency):
    def __init__(self, *s):
        self.urllist = list(s) # url 리스트
        self.cnt = len(list(s)) # url 보유 숫자
        
    def getWordsFrequency(self, reverse = False):
        #이 부분은 이전 과제와 거의 유사, 정렬 부분에 변화 O
        self.dic = dict() #사전 생성
        for k in range(self.cnt):
            source = urllib2.urlopen(self.urllist[k]).read()

            for i in range(0, source.count('<!--')):
                if source.find("<!--") != -1:
                    source = source.replace(source[source.find("<!--"):source.find("-->")+3:],"")

            for i in range(0,source.count("<")):
                if source.find("<") != -1:
                    source = source.replace(source[source.find("<"):source.find(">")+1:],"")
                    
            words = source.split()
            s = ' '.join(words)
            
            #구두문자 제거
            words = list(s)
            for i in range(len(words)):    
                for k in range(len(string.punctuation)):
                    if(words[i] == string.punctuation[k]):
                        words[i] = ' '
            words = ''.join(words)

            #구두문자가 제거된 문자열을 공백을 제외한 리스트로 만들기
            words = words.split()
            
            #영어 불용어 제거
            eng = ['a', 'and', 'are', 'as', 'at', 'be', 'but', 'by', 'for', 'if', 'in', 'into', 'is', 'it', 'no', 'not', 'of', 'on', 'or', 's', 'such', 't', 'that', 'the', 'their', 'then', 'there', 'these', 'they', 'this', 'to', 'was', 'will', 'with']
            for i in range(len(words)):
                for k in range(len(eng)):
                    if words[i] == eng[k]:
                        words[i] = ' '
            words = ' '.join(words)

            #영어 불용어가 제거된 문자열을 공백을 제외한 리스트로 만들기
            words = words.split()
            
            #사전 만들고 카운트 세기
            for word in words:
                if word not in self.dic:
                    self.dic[word] = 1
                else:
                    self.dic[word] += 1 
                    
            # 정렬하기 -> 변화 O 부분
            if (reverse == False): #내림차순!
                self.sortlist = sorted(self.dic.iteritems(), key=operator.itemgetter(1), reverse=True) #역순이므로 reverse를 True로 바꾸어 줌.
            else: #오름차순!
                self.sortlist = sorted(self.dic.iteritems(), key=operator.itemgetter(1))# 역순이 아니므로 default값 False
            return self.sortlist
        

w4 = OrderedWebWordsFrequency('http://www.times.com', 'https://www.amazon.com', 'https://github.com') 
print w4.getWordsFrequency()
print
print 
print w4.getWordsFrequency(reverse=True)

[('class', 903), ('div', 558), ('li', 509), ('story', 462), ('data', 436), ('com', 435), ('http', 424), ('nytimes', 351), ('href', 343), ('h2', 320), ('www', 319), ('article', 288), ('html', 253), ('id', 252), ('2015', 216), ('12', 204), ('heading', 199), ('collection', 179), ('section', 167), ('renderstyle', 129), ('rank', 127), ('column', 121), ('ul', 110), ('0', 106), ('index', 105), ('nyt', 102), ('04', 98), ('theme', 94), ('pages', 94), ('i', 93), ('h3', 92), ('1', 83), ('HpHedThumbWell', 81), ('summary', 78), ('p', 77), ('icon', 74), ('e', 72), ('span', 71), ('button', 69), ('region', 66), ('images', 66), ('ad', 66), ('src', 65), ('n', 64), ('link', 61), ('static01', 60), ('h6', 59), ('layout', 55), ('opinion', 55), ('img', 52), ('thumb', 52), ('hpHeader', 52), ('alt', 49), ('top', 49), ('function', 47), ('The', 46), ('2', 45), ('jpg', 45), ('hidden', 44), ('content', 44), ('font', 40), ('timestamp', 39), ('script', 38), ('topnews', 37), ('03', 35), ('video', 35), ('raquo', 34), 

#[15번 문제]
다음과 같은 코딩이 가능하도록 OrderedWebWordsFrequency 안에 반복자와 관련된 메소드를 추가하시오.

    >>> for i in w4:
    >>>       print i
    ('site, 12)
    ('science', 11)
    ('hello', 8)
    ('world', 2)
    ('program', 1)
    ('python', 1)

=> 14번, 15번 문제를 좀 더 보기 편하게 하기 위해 OrderedWebWordsFrequency2로 하였습니다. 

=> 이 소스 아래에 문제에서 원하는대로 OrderedWebWordsFrequency안에 넣은 소스를 따로 작성하였습니다.

In [14]:
class OrderedWebWordsFrequency2(OrderedWebWordsFrequency):    
    def __init__(self, *s):
        self.urllist = list(s) # url 리스트
        self.cnt = len(list(s)) # url 보유 숫자
        self.getWordsFrequency()
     
    def __iter__(self):
        self.p = 0 #읽어온 인덱스 확인 변수
        return self
    
    def next(self):
        if (self.p >= len(self.sortlist)): #읽은 인덱스와 총 길이 비교하여 
            raise StopIteration     # 읽을 수 없으면(범위를 넘으면) 예외 발생   
        line = self.sortlist[self.p] #현재읽어야하는 부분을 넘겨준다.
        self.p += 1     #인덱스 수 증가
        return line                 # 읽은 라인을 리턴한다.

w5 = OrderedWebWordsFrequency2('http://www.times.com', 'https://www.amazon.com', 'https://github.com')
for i in w5:
    print i

('class', 902)
('div', 556)
('li', 509)
('story', 463)
('data', 436)
('com', 435)
('http', 424)
('nytimes', 351)
('href', 343)
('h2', 320)
('www', 319)
('article', 288)
('html', 253)
('id', 252)
('2015', 217)
('12', 205)
('heading', 199)
('collection', 179)
('section', 167)
('renderstyle', 129)
('rank', 127)
('column', 121)
('ul', 110)
('0', 106)
('index', 105)
('nyt', 102)
('04', 101)
('theme', 94)
('pages', 94)
('i', 93)
('h3', 92)
('1', 83)
('HpHedThumbWell', 81)
('summary', 78)
('p', 77)
('icon', 74)
('e', 72)
('span', 71)
('button', 69)
('region', 66)
('images', 66)
('ad', 66)
('src', 65)
('n', 64)
('link', 61)
('static01', 60)
('h6', 57)
('layout', 55)
('opinion', 55)
('img', 52)
('thumb', 52)
('hpHeader', 52)
('alt', 49)
('top', 49)
('function', 47)
('jpg', 46)
('2', 45)
('hidden', 44)
('content', 44)
('The', 43)
('font', 40)
('timestamp', 39)
('script', 38)
('topnews', 37)
('03', 35)
('video', 35)
('raquo', 34)
('o', 33)
('nocontent', 32)
('meta', 32)
('news', 31)
('name', 31)


In [13]:
class OrderedWebWordsFrequency(WebWordsFrequency):
    def __init__(self, *s):
        self.urllist = list(s) # url 리스트
        self.cnt = len(list(s)) # url 보유 숫자
        self.getWordsFrequency()
        
    def getWordsFrequency(self, reverse = False):
        self.dic = dict() #사전 생성
        for k in range(self.cnt):
            source = urllib2.urlopen(self.urllist[k]).read()

            for i in range(0, source.count('<!--')):
                if source.find("<!--") != -1:
                    source = source.replace(source[source.find("<!--"):source.find("-->")+3:],"")

            for i in range(0,source.count("<")):
                if source.find("<") != -1:
                    source = source.replace(source[source.find("<"):source.find(">")+1:],"")
                    
            words = source.split()
            s = ' '.join(words)
            
            #구두문자 제거
            words = list(s)
            for i in range(len(words)):    
                for k in range(len(string.punctuation)):
                    if(words[i] == string.punctuation[k]):
                        words[i] = ' '
            words = ''.join(words)

            #구두문자가 제거된 문자열을 공백을 제외한 리스트로 만들기
            words = words.split()
            
            #영어 불용어 제거
            eng = ['a', 'and', 'are', 'as', 'at', 'be', 'but', 'by', 'for', 'if', 'in', 'into', 'is', 'it', 'no', 'not', 'of', 'on', 'or', 's', 'such', 't', 'that', 'the', 'their', 'then', 'there', 'these', 'they', 'this', 'to', 'was', 'will', 'with']
            for i in range(len(words)):
                for k in range(len(eng)):
                    if words[i] == eng[k]:
                        words[i] = ' '
            words = ' '.join(words)

            #영어 불용어가 제거된 문자열을 공백을 제외한 리스트로 만들기
            words = words.split()
            
            #사전 만들고 카운트 세기
            for word in words:
                if word not in self.dic:
                    self.dic[word] = 1
                else:
                    self.dic[word] += 1 
                    
            #정렬하기
            if (reverse == False): #내림차순!
                self.sortlist = sorted(self.dic.iteritems(), key=operator.itemgetter(1), reverse=True)
            else: #오름차순!
                self.sortlist = sorted(self.dic.iteritems(), key=operator.itemgetter(1))
            return self.sortlist
        
    def __iter__(self):
        self.p = 0
        return self
    
    def next(self):
        if (self.p >= len(self.sortlist)): 
            raise StopIteration     # 읽을 수 없으면 예외 발생   
        line = self.sortlist[self.p]
        self.p += 1     
        return line                 # 읽은 라인을 리턴한다.

w4 = OrderedWebWordsFrequency('http://www.times.com', 'https://www.amazon.com', 'https://github.com') 
w4.getWordsFrequency()
for i in w4:
    print i

('class', 902)
('div', 556)
('li', 509)
('story', 463)
('data', 436)
('com', 435)
('http', 424)
('nytimes', 351)
('href', 343)
('h2', 320)
('www', 319)
('article', 288)
('html', 253)
('id', 252)
('2015', 217)
('12', 205)
('heading', 199)
('collection', 179)
('section', 167)
('renderstyle', 129)
('rank', 127)
('column', 121)
('ul', 110)
('0', 106)
('index', 105)
('nyt', 102)
('04', 101)
('theme', 94)
('pages', 94)
('i', 93)
('h3', 92)
('1', 83)
('HpHedThumbWell', 81)
('summary', 78)
('p', 77)
('icon', 74)
('e', 72)
('span', 71)
('button', 69)
('region', 66)
('images', 66)
('ad', 66)
('src', 65)
('n', 64)
('link', 61)
('static01', 60)
('h6', 57)
('layout', 55)
('opinion', 55)
('img', 52)
('thumb', 52)
('hpHeader', 52)
('alt', 49)
('top', 49)
('function', 47)
('jpg', 46)
('2', 45)
('hidden', 44)
('content', 44)
('The', 43)
('font', 40)
('timestamp', 39)
('script', 38)
('topnews', 37)
('03', 35)
('video', 35)
('raquo', 34)
('o', 33)
('nocontent', 32)
('meta', 32)
('news', 31)
('name', 31)
